enhancements/mco-kargs-day1-support.md

```diff
@@ -0,0 +1,344 @@
+  - "@runcom"
+
+creation-date: 2019-09-12
+last-updated: 2019-09-16
+status: **provisional**|implementable|implemented|deferred|rejected|withdrawn|replaced
+---
+
+# Kernel Args - Day 1 Support
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement proposes adding the ability for admins to configure custom
+kernel arguments during initial RHCOS node configuration. With this enhancement,
+both OS upgrade to the latest [machine-os-content](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md#os-updates)
+and custom kernel args can be applied to a node during the first boot of the
+machine, followed by a single reboot.
+
+## Motivation
+
+Kernel arguments for RHCOS can be set today using a MachineConfig as a
+[`day-2` operation](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfiguration.md#kernelarguments),
+i.e. after cluster install, incurring an additional reboot to apply. Even if the
+kernel args are configured at [installation time](https://github.com/openshift/installer/blob/master/docs/user/customization.md#install-time-customization-for-machine-configuration),
+the node will still incur a double reboot to apply the args. Furthermore, that
+reboot will not be coordinated by the [MachineConfigController](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigController.md#updatecontroller),
+resulting in possible cluster downtime.
+
+We can provide a better experience by allowing kernel arguments to be set as a
+`day-1` operation, when a node is being initialized with the current OS and
+cluster version.
+
+
+### Goals
+
+- [ ] Cluster admins can define custom kernel arguments for RHCOS nodes during
+  cluster installation, without rebooting the machine twice.
+- [ ] kernel args
+  can be configured via MachineConfigs and will be applied to new nodes joining
+  the cluster, without inclurring an extra reboot.
+- [ ] In clusters installed using the 4.3 installer or higher, nodes with kernel
+  arguments configured via MachineConfigs have these kernel args applied with
+  the update without incurring an extra reboot.
+
+### Non-Goals
+
+- Does not support defining custom kernel arguments on non-RHCOS machines.
+- Does not support applying kernel args on first boot for nodes in clusters
+  installed using a 4.1 or earlier installer.
+- Does not protect against setting kernel arguments that may affect the
+  performance of the node or cluster.
+- Does not take into account kernel args set manually on the node when applying
+  kernel args from the MachineConfigs. Any kernel args manually set or unset
+  will be overwritten on the next update.
+
+## Proposal
+
+#### Part 1
```

Comment:
@abhinavdahiya are we ok with this making it into install-config or should it be integrated into manifest manipulation?

Reply:
Right, it should Just Work to pass kargs via [additional manifests](https://github.com/openshift/installer/blob/master/docs/user/customization.md#install-time-customization-for-machine-configuration).

Reply:
Side note, as you all know I have a wrapper around the installer with improved (IMO) ergonomics, and handling additional manifests conveniently is part of that: https://github.com/cgwalters/xokdinst/commit/0bef3d726af5fa76fbc19f35735757494808ee43

Reply:
I think if we pull this off we should also deprecate the `hyperthreading` install-config flag, and point people to setting `nosmt` directly via a `MachineConfig` passed in manifests.

Reply:
I think we should start with manifest manipulation only for general kernel arguments.

And then make a case for options that benefits our users. For example, hyperthreading is something _apparently_ a lot of our users care about and therefore it was added to the install-config as option.

Similarly we could think about maybe a graduation flow.

Reply:
I am ok with going with machineconfigs in manifests at first, but I'm not sure about long term given that it's not a supported by the installer: https://github.com/openshift/installer/blame/master/docs/user/customization.md#L224-L225

Reply:
"This can affect upgradability of your cluster as the `machine-config-operator` can mark clusters tainted when user defined [MachineConfig][machine-config] objects are present in the cluster."

is...at best poorly worded.  I mean, we certainly support user-defined MachineConfig objects.  It's clear it can affect upgradability but...we support it, right?